### DISCUSSION TITLE
Module.Load often results in FileNotFoundException

### DISCUSSION THREAD

**Discussion by rouke-broersma** [jbevain/cecil#745] [Q&A] [answered]
Hi 👋 

In stryker we use Mono.Cecil to load embedded resources from dll's. However loading the dll often fails with a FileNotFoundException even though I can confirm on the filesystem that the dll does exist in that location. We get the output location directly from MSBuild. Can anyone give some hints as to why it might be that Cecil cannot open the dll even though it is there?

Here is the class where we load the Embedded Resources: https://github.com/stryker-mutator/stryker-net/blob/v1.0/src/Stryker.Core/Stryker.Core/Initialisation/EmbeddedResourcesGenerator.cs

**Comment by jbevain** [comment, marked as answer]
Hello,

This is not something that I ever seen. Cecil doesn't throw `FileNotFoundException` by itself. It can sometimes throw a `AssemblyResolutionException` which extends `FileNotFoundException`.

If an actual `FileNotFoundException` it's being thrown by .NET as the OS is telling the .NET framework the file asked for the path doesn't exist.

**Reply by rouke-broersma** [reply]
Example: https://github.com/stryker-mutator/stryker-net/issues/1195

As the user says in the issue the dll was found on the filesystem manually so it did exist.

- Our process is as follows:  
- Use msbuild to build the project  
- Use Buildalyzer (msbuild 'scanner' library) to get project info and sources from msbuild  
- Use Cecil to load embedded resources from dll into memory
...
...    
- Use Roslyn SDK to compile sources with embedded resources  

Do you think it's possible that Cecil cannot access the dll, even though it exists on disk, because it's still locked by msbuild? The reason I'm asking is because so far I have been unable to reproduce this in a debug session, only seems to happen with the published version.

**Reply by rouke-broersma** [reply]
On second thought I think there are more specific errors when a file is locked or the current user does not have permission.

**Reply by jbevain** [reply]
If the file was locked it would be another kind of exception, no?

Down the line that's the code opening the file stream:

https://github.com/jbevain/cecil/blob/3c4ea3f8f4fcb688194f5f1e501ab01642d4bdf4/Mono.Cecil/ModuleDefinition.cs#L1098

It's telling the OS it's ok to Share a read handle.

One thing I'm not sure I understand is that you're saying that:

> Use Cecil to load embedded resources from dll into memory

But the stacktrace is clearly trying to open a filestream from the disk, not from memory.

**Reply by rouke-broersma** [reply]
Sorry I meant to say we load the dll from disk so we can copy the embedded resources from the dll into memory. This so we later can do an in-memory roslyn compilation with embedded resources.

**Reply by jbevain** [reply]
I really can't help much I'm afraid, .NET doesn't throw `FileNotFoundException`s if the file is there :)

It sounds like an issue related to timing possibly, where at the time you try to get the file MSBuild hasn't copied or fully copied the DLL where you expect it to.

**Reply by SimonCropp** [reply]
@rouke-broersma i recommend placing a File.Exists check [here](https://github.com/stryker-mutator/stryker-net/blob/v1.0/src/Stryker.Core/Stryker.Core/Initialisation/EmbeddedResourcesGenerator.cs#L35) with a throw if not.  ensure that your assumption is correct.

**Reply by rouke-broersma** [reply]
Yep that's a good idea!
